src/java.base/share/classes/java/lang/reflect/code/op/CoreOp.java

```diff
@@ -3676,50 +3729,86 @@ public static InvokeOp invoke(MethodRef invokeDescriptor, Value... args) {
+     * <p>
+     * The invoke return type is the invoke descriptors return type.
      *
-     * @param invokeDescriptor the invocation descriptor
-     * @param args             the invoke parameters
+     * @param invokeDescriptor the invoke descriptor
+     * @param args             the invoke arguments
      * @return the invoke operation
      */
     public static InvokeOp invoke(MethodRef invokeDescriptor, List<Value> args) {
         return invoke(invokeDescriptor.type().returnType(), invokeDescriptor, args);
     }
 
     /**
-     * Creates an invoke instance or class (static) operation.
+     * Creates an invoke operation modeling an invocation to an
+     * instance or static (class) method with no variable arguments.
+     * <p>
+     * The invoke kind of the invoke operation is determined by
+     * comparing the argument count with the invoke descriptor's
+     * parameter count. If they are equal then the invoke kind is
+     * {@link InvokeOp.InvokeKind#STATIC static}. If the parameter count
+     * plus one is equal to the argument count then the invoke kind
+     * is {@link InvokeOp.InvokeKind#STATIC instance}.
      *
-     * @param returnType       the invocation return type
-     * @param invokeDescriptor the invocation descriptor
-     * @param args             the invoke parameters
+     * @param returnType       the invoke return type
+     * @param invokeDescriptor the invoke descriptor
+     * @param args             the invoke arguments
      * @return the invoke operation
      */
     public static InvokeOp invoke(TypeElement returnType, MethodRef invokeDescriptor, Value... args) {
         return invoke(returnType, invokeDescriptor, List.of(args));
     }
 
     /**
-     * Creates an invoke instance or class (static) operation.
+     * Creates an invoke operation modeling an invocation to an
+     * instance or static (class) method with no variable arguments.
+     * <p>
+     * The invoke kind of the invoke operation is determined by
+     * comparing the argument count with the invoke descriptor's
+     * parameter count. If they are equal then the invoke kind is
+     * {@link InvokeOp.InvokeKind#STATIC static}. If the parameter count
+     * plus one is equal to the argument count then the invoke kind
+     * is {@link InvokeOp.InvokeKind#STATIC instance}.
      *
-     * @param returnType       the invocation return type
-     * @param invokeDescriptor the invocation descriptor
-     * @param args             the invoke parameters
+     * @param returnType       the invoke return type
+     * @param invokeDescriptor the invoke descriptor
+     * @param args             the invoke arguments
      * @return the invoke super operation
      */
     public static InvokeOp invoke(TypeElement returnType, MethodRef invokeDescriptor, List<Value> args) {
-        return new InvokeOp.InvokeInstanceClassOp(returnType, invokeDescriptor, args);
+        int paramCount = invokeDescriptor.type().parameterTypes().size();
```

Comment:
I wonder if, for clients, it would be easier to use a descriptor for the non-varargs part, then specify a TypeElement for the varargs type, and let the API figure out how to wield those together. In a way the fact that "varargs are methods whose last parameter is an array" seems an impl detail?

Reply:
Yeah - the fact that a `MethodRef` is basically bound to a symbolic reference is quite constraining here, as implementation details (varargs are arrays) end up leaking in the method ref. But that seems to be unavoidable to some extent.